File: tests_cpp/test_tree.cpp
#include <cmath>
#include <gtest/gtest.h>
#include <iostream>
#include <vector>
#include <velesquant/volatility/c_tree.h>
#include <velesquant/models/utility.h> // For OptionType

using namespace velesquant;

// Helper to generate flat curves
std::vector<double> generateFlatCurve(const std::vector<double> &times,
                                      double value) {
  std::vector<double> vals;
  vals.reserve(times.size());
  for (size_t i = 0; i < times.size(); ++i) {
    vals.push_back(value);
  }
  return vals;
}

// Helper to calculate Forwards
std::vector<double> calculateForwards(double spot, double r, double q,
                                      const std::vector<double> &times) {
  std::vector<double> fwds;
  fwds.reserve(times.size());
  for (double t : times) {
    fwds.push_back(spot * std::exp((r - q) * t));
  }
  return fwds;
}

class CTreeTestFixture : public ::testing::Test {
protected:
  double spot = 100.0;
  double r = 0.05;
  double q = 0.02;
  double sigma = 0.2;
  double T = 1.0;

  std::vector<double> timeGrid;
  std::vector<double> vols;
  std::vector<double> forwards;
  std::vector<double> rates;
  std::vector<double> divs;

  void SetUp() override {
    // Grid should cover the option maturity T=1.0
    // Let's define points at 0, 0.5, 1.0, 1.5, 2.0
    timeGrid = {0.0, 0.5, 1.0, 1.5, 2.0};
    vols = generateFlatCurve(timeGrid, sigma);
    forwards = calculateForwards(spot, r, q, timeGrid);

    // Pass r and q as flat curves (as CTree currently only uses first element
    // anyway)
    rates = generateFlatCurve(timeGrid, r);
    divs = generateFlatCurve(timeGrid, q);
  }
};

TEST_F(CTreeTestFixture, Construction) {
  CTree tree(spot, timeGrid, forwards, vols, rates, divs);
  SUCCEED();
}

TEST_F(CTreeTestFixture, EuropeanOptionPricing) {
  CTree tree(spot, timeGrid, forwards, vols, rates, divs);
  double strike = 100.0;
  int steps = 50;

  // calculateBinomial(strike, Maturity, Nnodes, style, pay, treeType)

  double priceBinomial = tree.calculateBinomial(
      strike, T, steps, exStyle::European, OptionType::Call, tType::recomb);

  // Analytical Black-Scholes estimate: ~9.23
  EXPECT_NEAR(priceBinomial, 9.23, 0.5);

  double priceTrinomial = tree.calculateTrinomial(
      strike, T, steps, exStyle::European, OptionType::Call, tType::recomb);

  EXPECT_NEAR(priceTrinomial, 9.23, 0.5);
}

TEST_F(CTreeTestFixture, AmericanOptionPricing) {
  CTree tree(spot, timeGrid, forwards, vols, rates, divs);
  double strike = 100.0;
  int steps = 50;

  double putEuro = tree.calculateBinomial(strike, T, steps, exStyle::European,
                                          OptionType::Put, tType::recomb);

  double putAmer = tree.calculateBinomial(strike, T, steps, exStyle::American,
                                          OptionType::Put, tType::recomb);

  // American Put >= European Put
  EXPECT_GE(putAmer, putEuro - 1e-10);
  EXPECT_GT(putAmer, 0.0);
}

File: tests/models/test_tree.py
"""Tests for CTree (Binomial/Trinomial Tree) model."""

from velesquant import native
from velesquant.models import TreeModel


def test_ctree_binding():
    """Test that CTree can be instantiated and called."""
    # Market data: spot, times, forwards, implied vols
    S = 100.0
    T = [0.5, 1.0]
    F = [100.5, 101.0]
    IV = [0.2, 0.22]
    r = [0.01, 0.01]
    q = [0.002, 0.002]

    tree = native.CTree(S, T, F, IV, r, q)
    # European call using binomial tree
    price = tree.calculate_binomial(
        100.0,
        1.0,
        100,
        native.ExerciseStyle.European,
        native.OptionType.Call,
        native.TreeType.Recombining,
    )
    assert isinstance(price, float)
    assert price >= 0.0


def test_ctree_american_put():
    """Test American put pricing - should be >= European put."""
    S = 100.0
    T = [0.5, 1.0]
    F = [100.5, 101.0]
    IV = [0.2, 0.22]

    tree = native.CTree(S, T, F, IV)

    euro_put = tree.calculate_binomial(
        100.0,
        1.0,
        100,
        native.ExerciseStyle.European,
        native.OptionType.Put,
        native.TreeType.Recombining,
    )
    amer_put = tree.calculate_binomial(
        100.0,
        1.0,
        100,
        native.ExerciseStyle.American,
        native.OptionType.Put,
        native.TreeType.Recombining,
    )

    # American put should be >= European put
    assert amer_put >= euro_put


def test_ctree_trinomial():
    """Test trinomial tree pricing."""
    S = 100.0
    T = [0.5, 1.0]
    F = [100.5, 101.0]
    IV = [0.2, 0.22]

    tree = native.CTree(S, T, F, IV)
    price = tree.calculate_trinomial(
        100.0,
        1.0,
        50,
        native.ExerciseStyle.European,
        native.OptionType.Call,
        native.TreeType.Recombining,
    )
    assert isinstance(price, float)
    assert price >= 0.0


def test_trees_wrapper_init():
    model = TreeModel(100.0, [0.5, 1.0], [102.0, 105.0], [0.2, 0.2])
    price = model.price_binomial(105.0, 1.0, n_nodes=50)
    assert price > 0

    price_tri = model.price_trinomial(105.0, 1.0, n_nodes=50)
    assert price_tri > 0

    d = model.to_dict()
    assert d["type"] == "TreeModel"
    assert d["spot"] == 100.0
